libafl_frida/src/asan_errors.rs

```diff
@@ -39,10 +44,80 @@ pub(crate) enum AsanError {
     WriteAfterFree(AsanReadWriteError),
     DoubleFree((usize, AllocationMetadata, Backtrace)),
     UnallocatedFree((usize, Backtrace)),
-    Unknown(([usize; 32], usize, (u16, u16, usize, usize), Backtrace)),
+    Unknown(
+        (
+            [usize; ASAN_SAVE_REGISTER_COUNT],
+            usize,
+            (u16, u16, usize, usize),
+            Backtrace,
+        ),
+    ),
     Leak((usize, AllocationMetadata)),
-    StackOobRead(([usize; 32], usize, (u16, u16, usize, usize), Backtrace)),
-    StackOobWrite(([usize; 32], usize, (u16, u16, usize, usize), Backtrace)),
+    StackOobRead(
+        (
+            [usize; ASAN_SAVE_REGISTER_COUNT],
+            usize,
+            (u16, u16, usize, usize),
+            Backtrace,
+        ),
+    ),
+    StackOobWrite(
+        (
+            [usize; ASAN_SAVE_REGISTER_COUNT],
+            usize,
+            (u16, u16, usize, usize),
+            Backtrace,
+        ),
+    ),
+    BadFuncArgRead((String, usize, usize, usize, Backtrace)),
+    BadFuncArgWrite((String, usize, usize, usize, Backtrace)),
+}
+
+#[cfg(target_arch = "x86_64")]
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub(crate) struct AsanReadWriteError {
+    pub registers: [usize; ASAN_SAVE_REGISTER_COUNT],
+    pub pc: usize,
+    pub fault: (Option<u16>, Option<u16>, usize, usize),
+    pub metadata: AllocationMetadata,
+    pub backtrace: Backtrace,
+}
+
+#[cfg(target_arch = "x86_64")]
+#[allow(clippy::type_complexity)]
+#[derive(Debug, Clone, Serialize, Deserialize, SerdeAny)]
+pub(crate) enum AsanError {
+    OobRead(AsanReadWriteError),
+    OobWrite(AsanReadWriteError),
+    ReadAfterFree(AsanReadWriteError),
+    WriteAfterFree(AsanReadWriteError),
+    DoubleFree((usize, AllocationMetadata, Backtrace)),
+    UnallocatedFree((usize, Backtrace)),
+    Unknown(
+        (
+            [usize; ASAN_SAVE_REGISTER_COUNT],
+            usize,
+            (Option<u16>, Option<u16>, usize, usize),
```

Comment:
Why do we need another copy of AsanError? Are the x86_64 and aarch64 really that different?

Reply:
I mean I would like to change the AsanError to have `basereg` and `indexreg` wrapped by Option, instead of making a copy

Instructions like `mov rdi, [rip + 0xaaaa]`, does not have an indexreg.
Currently what we are doing is 
https://github.com/AFLplusplus/LibAFL/blob/2055eabede2316771a788b43ba135dcbe5a41127/libafl_frida/src/asan_rt.rs#L1938
set indexreg to the maximum of u16 value, (to indicate indexreg is not valid.)
I propose to use Some/None here.

Reply:
works for me. Go ahead and make the changes.

Reply:
I made the change (but I didn't (can't) test compiling on aarch64_